File: java/src/main/java/com/mandiri/qris/service/MandiriQrisService.java
package com.mandiri.qris.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mandiri.qris.config.MandiriQrisConfig;
import com.mandiri.qris.dto.CreateQrisRequest;
import com.mandiri.qris.dto.QrisResponse;
import com.mandiri.qris.dto.PaymentStatusResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Slf4j
@Service
@RequiredArgsConstructor
public class MandiriQrisService {
    
    private final MandiriQrisConfig config;
    private final RestTemplate restTemplate;
    private final ObjectMapper objectMapper;
    
    private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
    
    /**
     * Get access token (with caching)
     */
    @Cacheable(value = "mandiriTokenCache", key = "'token'")
    public String getAccessToken() {
        try {
            String timestamp = getCurrentTimestamp();
            String signature = generateSignature(config.getClientId() + "|" + timestamp, config.getClientSecret());
            
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.set("X-TIMESTAMP", timestamp);
            headers.set("X-CLIENT-KEY", config.getClientId());
            headers.set("X-SIGNATURE", signature);
            
            Map<String, String> body = new HashMap<>();
            body.put("grantType", "client_credentials");
            
            HttpEntity<Map<String, String>> request = new HttpEntity<>(body, headers);
            
            ResponseEntity<String> response = restTemplate.postForEntity(
                config.getBaseUrl() + "/openapi/auth/v2.0/access-token/b2b",
                request,
                String.class
            );
            
            if (response.getStatusCode() == HttpStatus.OK) {
                JsonNode jsonNode = objectMapper.readTree(response.getBody());
                return jsonNode.get("accessToken").asText();
            } else {
                throw new RuntimeException("Failed to get access token: HTTP " + response.getStatusCode());
            }
            
        } catch (Exception e) {
            log.error("Failed to get access token", e);
            throw new RuntimeException("Failed to get access token: " + e.getMessage(), e);
        }
    }
    
    /**
     * Create QRIS payment
     */
    public QrisResponse createQris(CreateQrisRequest request) {
        try {
            String token = getAccessToken();
            String timestamp = getCurrentTimestamp();
            String expiryTime = getExpiryTimestamp(config.getQrisExpiryMinutes());
            
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(token);
            headers.set("X-TIMESTAMP", timestamp);
            headers.set("X-PARTNER-ID", config.getMerchantNmid());
            headers.set("X-EXTERNAL-ID", request.getReference());
            
            Map<String, Object> body = new HashMap<>();
            body.put("partnerReferenceNo", request.getReference());
            
            Map<String, String> amount = new HashMap<>();
            amount.put("value", String.format("%.2f", request.getAmount()));
            amount.put("currency", "IDR");
            body.put("amount", amount);
            
            body.put("merchantId", config.getMerchantNmid());
            body.put("storeLabel", config.getMerchantName());
            body.put("terminalLabel", config.getMerchantCity());
            body.put("validityPeriod", expiryTime);
            
            if (request.getCallbackUrl() != null && !request.getCallbackUrl().isEmpty()) {
                Map<String, String> additionalInfo = new HashMap<>();
                additionalInfo.put("callbackUrl", request.getCallbackUrl());
                body.put("additionalInfo", additionalInfo);
            }
            
            HttpEntity<Map<String, Object>> httpRequest = new HttpEntity<>(body, headers);
            
            ResponseEntity<String> response = restTemplate.postForEntity(
                config.getBaseUrl() + "/openapi/qris/v1.0/qr-code-dynamic",
                httpRequest,
                String.class
            );
            
            if (response.getStatusCode() == HttpStatus.OK) {
                JsonNode jsonNode = objectMapper.readTree(response.getBody());
                String qrContent = jsonNode.get("qrContent").asText();
                String qrId = jsonNode.has("qrId") ? jsonNode.get("qrId").asText() : request.getReference();
                
                QrisResponse qrisResponse = new QrisResponse();
                qrisResponse.setQrId(qrId);
                qrisResponse.setQrString(qrContent);
                qrisResponse.setQrImageUrl("https://api.qrserver.com/v1/create-qr-code/?size=300x300&data=" + 
                    URLEncoder.encode(qrContent, StandardCharsets.UTF_8));
                qrisResponse.setStatus("PENDING");
                qrisResponse.setAmount(request.getAmount());
                qrisResponse.setReference(request.getReference());
                qrisResponse.setExpiredAt(expiryTime);
                
                return qrisResponse;
            } else {
                throw new RuntimeException("Failed to create QRIS: HTTP " + response.getStatusCode());
            }
            
        } catch (Exception e) {
            log.error("Failed to create QRIS", e);
            throw new RuntimeException("Failed to create QRIS: " + e.getMessage(), e);
        }
    }
    
    /**
     * Check payment status
     */
    public PaymentStatusResponse checkStatus(String qrId, String reference) {
        try {
            String token = getAccessToken();
            String timestamp = getCurrentTimestamp();
            
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(token);
            headers.set("X-TIMESTAMP", timestamp);
            headers.set("X-PARTNER-ID", config.getMerchantNmid());
            headers.set("X-EXTERNAL-ID", qrId);
            
            Map<String, String> body = new HashMap<>();
            body.put("originalPartnerReferenceNo", reference);
            body.put("originalReferenceNo", qrId);
            body.put("serviceCode", "47");
            
            HttpEntity<Map<String, String>> request = new HttpEntity<>(body, headers);
            
            ResponseEntity<String> response = restTemplate.postForEntity(
                config.getBaseUrl() + "/openapi/qris/v1.0/qr-code-dynamic/status",
                request,
                String.class
            );
            
            if (response.getStatusCode() == HttpStatus.OK) {
                JsonNode jsonNode = objectMapper.readTree(response.getBody());
                
                PaymentStatusResponse statusResponse = new PaymentStatusResponse();
                statusResponse.setQrId(qrId);
                
                if (jsonNode.has("transactionStatusCode")) {
                    String statusCode = jsonNode.get("transactionStatusCode").asText();
                    switch (statusCode) {
                        case "00":
                            statusResponse.setStatus("COMPLETED");
                            statusResponse.setPaidAt(jsonNode.has("transactionDate") ? 
                                jsonNode.get("transactionDate").asText() : getCurrentTimestamp());
                            statusResponse.setTransactionId(jsonNode.has("referenceNo") ? 
                                jsonNode.get("referenceNo").asText() : null);
                            break;
                        case "03":
                            statusResponse.setStatus("PENDING");
                            break;
                        case "05":
                            statusResponse.setStatus("EXPIRED");
                            break;
                        default:
                            statusResponse.setStatus("FAILED");
                            break;
                    }
                } else {
                    statusResponse.setStatus("UNKNOWN");
                }
                
                if (jsonNode.has("amount") && jsonNode.get("amount").has("value")) {
                    statusResponse.setAmount(jsonNode.get("amount").get("value").asDouble());
                }
                
                return statusResponse;
            } else {
                throw new RuntimeException("Failed to check status: HTTP " + response.getStatusCode());
            }
            
        } catch (Exception e) {
            log.error("Failed to check payment status", e);
            throw new RuntimeException("Failed to check status: " + e.getMessage(), e);
        }
    }
    
    /**
     * Generate HMAC SHA256 signature
     */
    private String generateSignature(String data, String secret) {
        try {
            Mac sha256Hmac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            sha256Hmac.init(secretKey);
            byte[] hash = sha256Hmac.doFinal(data.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(hash);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate signature", e);
        }
    }
    
    /**
     * Get current timestamp in ISO 8601 format
     */
    private String getCurrentTimestamp() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern(DATE_FORMAT));
    }
    
    /**
     * Get expiry timestamp
     */
    private String getExpiryTimestamp(int minutes) {
        return LocalDateTime.now().plusMinutes(minutes).format(DateTimeFormatter.ofPattern(DATE_FORMAT));
    }
}

File: aspnet/Services/MandiriQrisService.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Options;
using Newtonsoft.Json;

namespace MandiriQris.Services
{
    public class MandiriQrisService
    {
        private readonly MandiriQrisConfig _config;
        private readonly IMemoryCache _cache;
        private readonly HttpClient _httpClient;
        private readonly ILogger<MandiriQrisService> _logger;

        public MandiriQrisService(
            IOptions<MandiriQrisConfig> config, 
            IMemoryCache cache,
            HttpClient httpClient,
            ILogger<MandiriQrisService> logger)
        {
            _config = config.Value;
            _cache = cache;
            _httpClient = httpClient;
            _logger = logger;
            _httpClient.Timeout = TimeSpan.FromSeconds(_config.Timeout);
        }

        /// <summary>
        /// Get access token (with caching)
        /// </summary>
        public async Task<string> GetAccessTokenAsync()
        {
            var cacheKey = $"mandiri_qris_token_{_config.ClientId}";

            // Try to get from cache
            if (_cache.TryGetValue(cacheKey, out string? cachedToken) && cachedToken != null)
            {
                return cachedToken;
            }

            // Request new token
            var timestamp = DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'");
            var signature = GenerateSignature($"{_config.ClientId}|{timestamp}", _config.ClientSecret);

            var request = new HttpRequestMessage(HttpMethod.Post, $"{_config.BaseUrl}/openapi/auth/v2.0/access-token/b2b");
            request.Headers.Add("X-TIMESTAMP", timestamp);
            request.Headers.Add("X-CLIENT-KEY", _config.ClientId);
            request.Headers.Add("X-SIGNATURE", signature);

            var body = new { grantType = "client_credentials" };
            request.Content = new StringContent(JsonConvert.SerializeObject(body), Encoding.UTF8, "application/json");

            var response = await _httpClient.SendAsync(request);
            
            if (!response.IsSuccessStatusCode)
            {
                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogError("Failed to get access token: {Error}", errorContent);
                throw new Exception($"Failed to get access token: HTTP {response.StatusCode}");
            }

            var content = await response.Content.ReadAsStringAsync();
            var data = JsonConvert.DeserializeObject<dynamic>(content);

            if (data?.accessToken == null)
            {
                throw new Exception("Failed to get access token: Invalid response");
            }

            string token = data.accessToken;
            int expiresIn = data.expiresIn ?? 3600;

            // Cache token (with 60 seconds safety margin)
            _cache.Set(cacheKey, token, TimeSpan.FromSeconds(expiresIn - 60));

            return token;
        }

        /// <summary>
        /// Create QRIS payment
        /// </summary>
        public async Task<QrisResponse> CreateQrisAsync(CreateQrisRequest qrisRequest)
        {
            var token = await GetAccessTokenAsync();
            var timestamp = DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'");
            var expiryTime = DateTime.UtcNow.AddMinutes(_config.QrisExpiryMinutes).ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'");

            var request = new HttpRequestMessage(HttpMethod.Post, $"{_config.BaseUrl}/openapi/qris/v1.0/qr-code-dynamic");
            request.Headers.Add("Authorization", $"Bearer {token}");
            request.Headers.Add("X-TIMESTAMP", timestamp);
            request.Headers.Add("X-PARTNER-ID", _config.MerchantNmid);
            request.Headers.Add("X-EXTERNAL-ID", qrisRequest.Reference);

            var body = new
            {
                partnerReferenceNo = qrisRequest.Reference,
                amount = new
                {
                    value = qrisRequest.Amount.ToString("F2"),
                    currency = "IDR"
                },
                merchantId = _config.MerchantNmid,
                storeLabel = _config.MerchantName,
                terminalLabel = _config.MerchantCity,
                validityPeriod = expiryTime,
                additionalInfo = !string.IsNullOrEmpty(qrisRequest.CallbackUrl) 
                    ? new { callbackUrl = qrisRequest.CallbackUrl }
                    : null
            };

            request.Content = new StringContent(JsonConvert.SerializeObject(body), Encoding.UTF8, "application/json");

            var response = await _httpClient.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogError("Failed to create QRIS: {Error}", errorContent);
                throw new Exception($"Failed to create QRIS: HTTP {response.StatusCode}");
            }

            var content = await response.Content.ReadAsStringAsync();
            var data = JsonConvert.DeserializeObject<dynamic>(content);

            if (data?.qrContent == null)
            {
                throw new Exception("Failed to create QRIS: Invalid response");
            }

            string qrContent = data.qrContent;
            string qrId = data.qrId ?? qrisRequest.Reference;

            return new QrisResponse
            {
                QrId = qrId,
                QrString = qrContent,
                QrImageUrl = $"https://api.qrserver.com/v1/create-qr-code/?size=300x300&data={Uri.EscapeDataString(qrContent)}",
                Status = "PENDING",
                Amount = qrisRequest.Amount,
                Reference = qrisRequest.Reference,
                ExpiredAt = expiryTime
            };
        }

        /// <summary>
        /// Check payment status
        /// </summary>
        public async Task<PaymentStatusResponse> CheckStatusAsync(string qrId, string reference)
        {
            var token = await GetAccessTokenAsync();
            var timestamp = DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'");

            var request = new HttpRequestMessage(HttpMethod.Post, $"{_config.BaseUrl}/openapi/qris/v1.0/qr-code-dynamic/status");
            request.Headers.Add("Authorization", $"Bearer {token}");
            request.Headers.Add("X-TIMESTAMP", timestamp);
            request.Headers.Add("X-PARTNER-ID", _config.MerchantNmid);
            request.Headers.Add("X-EXTERNAL-ID", qrId);

            var body = new
            {
                originalPartnerReferenceNo = reference,
                originalReferenceNo = qrId,
                serviceCode = "47"
            };

            request.Content = new StringContent(JsonConvert.SerializeObject(body), Encoding.UTF8, "application/json");

            var response = await _httpClient.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogError("Failed to check status: {Error}", errorContent);
                throw new Exception($"Failed to check status: HTTP {response.StatusCode}");
            }

            var content = await response.Content.ReadAsStringAsync();
            var data = JsonConvert.DeserializeObject<dynamic>(content);

            var statusResponse = new PaymentStatusResponse
            {
                QrId = qrId,
                Status = "UNKNOWN"
            };

            if (data?.transactionStatusCode != null)
            {
                string statusCode = data.transactionStatusCode;
                statusResponse.Status = statusCode switch
                {
                    "00" => "COMPLETED",
                    "03" => "PENDING",
                    "05" => "EXPIRED",
                    _ => "FAILED"
                };

                if (statusResponse.Status == "COMPLETED")
                {
                    statusResponse.PaidAt = data.transactionDate ?? DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'");
                    statusResponse.TransactionId = data.referenceNo;
                }
            }

            if (data?.amount?.value != null)
            {
                statusResponse.Amount = (double)data.amount.value;
            }

            return statusResponse;
        }

        /// <summary>
        /// Clear cached token
        /// </summary>
        public void ClearToken()
        {
            var cacheKey = $"mandiri_qris_token_{_config.ClientId}";
            _cache.Remove(cacheKey);
        }

        /// <summary>
        /// Generate HMAC SHA256 signature
        /// </summary>
        private string GenerateSignature(string data, string secret)
        {
            using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
            var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
            return Convert.ToBase64String(hash);
        }
    }

    // DTOs
    public class CreateQrisRequest
    {
        public double Amount { get; set; }
        public string Reference { get; set; } = string.Empty;
        public string? CallbackUrl { get; set; }
    }

    public class QrisResponse
    {
        public string QrId { get; set; } = string.Empty;
        public string QrString { get; set; } = string.Empty;
        public string QrImageUrl { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public double Amount { get; set; }
        public string Reference { get; set; } = string.Empty;
        public string ExpiredAt { get; set; } = string.Empty;
    }

    public class PaymentStatusResponse
    {
        public string QrId { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public double? Amount { get; set; }
        public string? PaidAt { get; set; }
        public string? TransactionId { get; set; }
    }
}
